### PR TITLE
UM-331 Create Bus Model

### PR DESCRIPTION
**Pull request by iud-wtag**
## Short Description

[UM-331](https://welltravel.atlassian.net/browse/UM-331)
This PR introduce creating a Bus model

## Types of Changes
* Creating Bus model with validation.
* Creating testing for the bus model.

Screenshot 2023-04-03 at 4 38 52 PM

## Test
Screenshot 2023-04-03 at 4 33 25 PM

## Review
Please have a close look on this PR and kindly put you valuable comments.

### CODE REVIEW

File: app/models/bus.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Bus < ApplicationRecord
+  validates :bus_name, :bus_type, :bus_brand, :bus_capacity, presence:true
```
Comment (chm-wtag) [inline app/models/bus.rb]: IMHO, we do not need to repeat the model name for each attribute name. The model is already named as `Bus`, so the attributes are definitely bus-attributes, right? Repeating the model name each time is too wordy, don't you think?

File: db/migrate/20230403091245_create_buses.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  def change
+    create_table :buses do |t|
+      t.string :bus_name, null: false
+      t.string :bus_type, null: false
+      t.string :bus_brand, null: false
+      t.integer :bus_capacity, default:0, null: false
+      t.timestamps
+    end
+
+    add_index :buses, :bus_name, unique: true
```
Comment (chm-wtag) [inline db/migrate/20230403091245_create_buses.rb]: can't we do this while defining the column at the first place? 🤔

File: app/models/bus.rb (hunk 1 above)
Comment (chm-wtag) [inline app/models/bus.rb]: what are the possible values of `bus_type`? Are they some fixed/previously defined values?

Comment (chm-wtag) [inline app/models/bus.rb]: what do we represent by `bus_capacity`? Will it become 0 when there is no seat left?

File: db/seeds.rb [hunk 3]
```diff
@@ -10,3 +10,12 @@
     role: [:user, :admin].sample
   )
 end  
+
+5.times do
+  Bus.create(
+    bus_name: Faker::Vehicle.license_plate,
+    bus_type: ["AC","Non-AC"].sample,
```
Comment (chm-wtag) [inline db/seeds.rb]: please use single quote for strings, it is a convention that we follow

File: spec/factories/buses.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+require "faker"
```
Comment (chm-wtag) [inline spec/factories/buses.rb]: Please check if we can skip this line

File: spec/models/bus_spec.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+
+RSpec.describe Bus, type: :model do
```
Comment (chm-wtag) [inline spec/models/bus_spec.rb]: please check if we can skip this explicit mention of `type`

File: spec/models/bus_spec.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  describe "bus model validations" do
+    let!(:bus) { FactoryBot.create(:bus) }
```
Comment (chm-wtag) [inline spec/models/bus_spec.rb]: please study difference between `let` and `let!`

**Review by chm-wtag — COMMENTED** [review]
please reach out in case of any confusions, thank you 🎉

File: spec/factories/buses.rb (hunk 4 above)
Comment (iud-wtag) [reply spec/factories/buses.rb:1]: Sorry apu, I didn't find any solution where `Faker` gem worked without the `require "faker"`. Is there any possible way to skip that line, that i'm missing?

File: db/migrate/20230403091245_create_buses.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+class CreateBuses < ActiveRecord::Migration[7.0]
+  def change
+    create_table :buses do |t|
+      t.string :bus_name, null: false
+      t.string :bus_type, null: false
+      t.string :bus_brand, null: false
+      t.integer :bus_capacity, default:0, null: false
```
Comment (zma-wtag) [inline db/migrate/20230403091245_create_buses.rb]: `bus_capacity` related to seat in bus? is there any possiblity we defined `bus_capacity` 10 but we assigned more than 10 seat in that bus?

File: spec/factories/buses.rb (hunk 4 above)
Comment (iud-wtag) [reply spec/factories/buses.rb]: Shifted the faker to rails_helper.rb file and skip it from this file. Thanks!

File: spec/models/bus_spec.rb (hunk 5 above)
Comment (iud-wtag) [reply spec/models/bus_spec.rb]: removed the explicit mention of `type: :model`. Thanks!

File: app/models/bus.rb (hunk 1 above)
Comment (iud-wtag) [reply app/models/bus.rb]: Yes. renamed those.

File: db/migrate/20230403091245_create_buses.rb (hunk 2 above)
Comment (iud-wtag) [reply db/migrate/20230403091245_create_buses.rb]: IDK why but defining the column at the first place doesn't adding to schema. But now i added it in the first place also found out that it is best practice to add_index for each migration which speed up the query time.

File: app/models/bus.rb (hunk 1 above)
Comment (iud-wtag) [reply app/models/bus.rb]: Yes it is previously defined now as `['AC','NON_AC']`. Ans also changed the type into integer now and set the enum value as `enum typed: { AC: 0, NON_AC: 1 }`

Comment (iud-wtag) [reply app/models/bus.rb]: here `bus_capacity` is the total number of seat a bus have. It's not depend on seat booking.

File: spec/models/bus_spec.rb (hunk 6 above)
Comment (iud-wtag) [reply spec/models/bus_spec.rb]: I've studied &  understood the difference between `let` and `let!`. Here I have user `let!` because for comparing uniqueness of  bus.name  it has to store the previous instance before invocation. So thats why I used that.

File: db/migrate/20230403091245_create_buses.rb (hunk 7 above)
Comment (iud-wtag) [reply db/migrate/20230403091245_create_buses.rb]: Yes `bus_capacity` is related to `seat` in bus and i have handled that we cannot assigned more `seat` than `bus_capacity`
